### PR TITLE
Fix the `mila init` and `mila code` commands on windows & WSL

### LINKED ISSUE #19: milatools setup for Windows
Hello, 
I've been having some issues getting VS code's debugger to run on a compute node from my Windows laptop.

**How to reproduce the issue:** 
1. [Install Python 3 on Windows](https://apps.microsoft.com/store/detail/python-310/9PJPW5LDXLZ5)
2. Execute the following commands in a Windows PowerShell terminal:
```
pip install milatools
mila init
```
At least in my case, the `mila init` command hangs after displaying the `Your public key does not appear be registered on the cluster. Register it? [Y/n]` message.

**The workaround:**
First you need to successfully run `mila init` to complete your cluster SSH setup.
1. [Install Linux on Windows with WSL](https://docs.microsoft.com/en-us/windows/wsl/install)
2. [Install Ubuntu on Windows](https://apps.microsoft.com/store/detail/ubuntu/9PDXGNCFSCZV?hl=en-gb&gl=gb)
3. Launch an Ubuntu terminal from the Windows Start menu (on first launch this will trigger an OS user account setup procedure; proceed with the setup)
4. Install Anaconda in the Ubuntu app:
4.1 Download the Anaconda linux setup script from [here](https://www.anaconda.com/products/distribution). For example: `wget https://repo.anaconda.com/archive/Anaconda3-2022.05-Linux-x86_64.sh`
4.2 Install Anaconda. For example: `bash Anaconda3-2022.05-Linux-x86_64.sh`
Answer yes to the following question: Do you wish the installer to initialize Anaconda3 by running conda init? [yes|no]
4.3 Launch a new Ubuntu terminal from Windows Start menu and confirm that the default Python’s version is now >= 3.8 (this is a milatools requirement) by executing the following command: `python -V`
5. Install and initialize milatools (still in Ubuntu):
```
pip install milatools
mila init
```

Then you need to generate a public RSA key on your Windows host and add it to the list of authorized keys on the cluster:
1. If not already done, [install Python 3 on Windows](https://apps.microsoft.com/store/detail/python-310/9PJPW5LDXLZ5)
2. Open a Windows PowerShell terminal from the Windows Start menu and install milatools by executing the following command: `pip install milatools`
3. On Windows, installation of milatools will end on a warning such as: `WARNING: The script mila.exe is installed in 'C:\Users\[USERNAME]\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\Scripts' which is not on PATH.` Take in note the absolute path of milatools’ installation folder
4. Add  the absolute path of milatools’ installation folder to Windows’ PATH environment variable by following [this procedure](https://datatofish.com/add-python-to-windows-path/)
5. Open a new Windows PowerShell terminal from the Windows Start menu and execute the following command: `mila init`
5.1 Due to a bug, the `mila init` command will hang after displaying the following message: `Your public key does not appear be registered on the cluster. Register it? [Y/n]`
6. Open a new PowerShell terminal and execute the following command to print your public RSA key: `type $env:USERPROFILE\.ssh\id_rsa.pub`
7. Use the following command to open an SSH connection to the mila cluster: `ssh mila`
8. Append your public RSA key (see step 6 above) at the end of the `~/.ssh/authorized_keys` file on Mila’s cluster’s host

Setup is done!
Now you can use the following procedure to launch VS code using a compute node from Mila’s cluster as as development environment:

1. Install [Microsoft’s Remote SSH](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-ssh) VS Code plugin
2. On Mila login node, ask for an interactive session and keep a note of the compute node hostname. For example: `salloc --cpus-per-task 1 --gres gpu:0 --time=0-1:00:00`
3. Execute the following command in a Windows PowerShell to launch VS Code using the compute node from the previous step as as development environment: code --remote ssh-remote+[COMPUTE_NODE_HOSTNAME].server.mila.quebec [SRC_CODE_DIRECTORY_ON_CLUSTER] 

**Lead to a bugfix:**
The problem seems to come from [trying to execute the `ssh-copy-id` command](https://github.com/mila-iqia/milatools/blob/58f4aff06f5c2f4fec65ff6ba921dee1558b9829/milatools/commands.py#L154) on a Windows host (this command does not exist on Windows).

I found a [one liner](https://www.chrisjhart.com/Windows-10-ssh-copy-id/#copy-ssh-key-to-remote-linux-device) that is supposed to be Windows' equivalent of `ssh-copy-id`, but that also hanged on me and I haven't had time to investigate further yet.

I will try to find some bandwidth to come up with a bugfix PR in the next few days.
Hopefully the workaround above will save some people some time in the meanwhile.

### LINKED ISSUE #37: mila code /full/path/to/remote/code does not work properly when launched from WSL Ubuntu on Windows 11
### Make sure you can reproduce the issue with the latest version available

Steps to reproduce:

```
# Install wsl on win11
# run all commands from linux terminal:
pip install milatools --upgrade
mila init  # puts publickey on cluster
mila code /full/path/to/remote/codebase

> (default) jdv@lyndon:/home
> $ mila code /home/mila/v/vivianoj/code/gfneco
> (->) $ salloc
> salloc: --------------------------------------------------------------------------------------------------
> salloc: # Using default long-cpu partition (CPU-only)
> salloc: --------------------------------------------------------------------------------------------------
> salloc: Granted job allocation 2991990
> salloc: Waiting for resource configuration
> salloc: Nodes cn-f003 are ready for job
> (local) $ '/mnt/c/Users/Joseph Viviano/AppData/Local/Programs/Microsoft VS Code/bin/code' -nw --remote ssh-remote+cn-f003.server.mila.quebec /home/mila/v/vivianoj/code/gfneco
```

### Describe the bug
At this point, `vscode` will successfully launch, but the instance will be in a strange state:

Screenshot 2023-03-22 134246

A few things are notable here. The terminal instance is local. The folder is also local, and pointing to a folder which does not exist. `vscode` is successfully tunnelled into `WSL:Ubuntu` (bottom left corner).  I am not sure whether a fix is to launch `mila code` from the windows side, will try that next. 

### Desktop (please complete the following information):
- OS: Windows 11 w/ WSL Ubuntu 

```
(default) jdv@lyndon:/home
$ uname -a
Linux lyndon 5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

Thank you :)

### LINKED ISSUE #63: Impossible to "ssh-copy-id mila" in the "mila init" process on Windows 11
### Make sure you can reproduce the issue with the latest version available
```
pip install milatools --upgrade
[milatools command e.g. mila code ...]
```

### What command did you run?
`mila init`
Then i entered my username on the mila cluster as asked by the command prompt.
Then I answered to the question
```
+Host mila
+  HostName login.server.mila.quebec
+  User basile.terver
+  PreferredAuthentications publickey,keyboard-interactive
+  Port 2222
+  ServerAliveInterval 120
+  ServerAliveCountMax 5
+
+
+Host mila-cpu
+  User basile.terver
+  Port 2222
+  ForwardAgent yes
+  StrictHostKeyChecking no
+  LogLevel ERROR
+  UserKnownHostsFile /dev/null
+  RequestTTY force
+  ConnectTimeout 600
+  ServerAliveInterval 120
+  ProxyCommand ssh mila "/cvmfs/config.mila.quebec/scripts/milatools/slurm-proxy.sh mila-cpu --mem=8G"
+  RemoteCommand /cvmfs/config.mila.quebec/scripts/milatools/entrypoint.sh mila-cpu
+
+
+Host *.server.mila.quebec !*login.server.mila.quebec
+  HostName %h
+  User basile.terver
+  ProxyJump mila
+
?
Is this OK?
```
Then I answered yes to the question 
`? You have no public keys. Generate one?`
Then I answered yes to the question
`? Your public key does not appear be registered on the cluster. Register it? Yes`
Then I got the error below.
### Describe the bug
In the "mila init process" on my Windows PC, I get stucked at this point, although I upgrade the milatools package.
```
Traceback (most recent call last):
  File "C:\Users\terve\anaconda3\envs\mila\Lib\site-packages\milatools\cli\commands.py", line 43, in main
    auto_cli(milatools)
  File "C:\Users\terve\anaconda3\envs\mila\Lib\site-packages\coleo\cli.py", line 656, in auto_cli
    result = run_cli(entry, args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\terve\anaconda3\envs\mila\Lib\site-packages\coleo\cli.py", line 628, in run_cli
    return call(opts=opts, args=args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\terve\anaconda3\envs\mila\Lib\site-packages\coleo\cli.py", line 587, in thunk
    result = fn(*args)
             ^^^^^^^^^
  File "C:\Users\terve\anaconda3\envs\mila\Lib\site-packages\milatools\cli\commands.py", line 161, in init
    here.run("ssh-copy-id", "mila")
  File "C:\Users\terve\anaconda3\envs\mila\Lib\site-packages\milatools\cli\local.py", line 28, in run
    return subprocess.run(
           ^^^^^^^^^^^^^^^
  File "C:\Users\terve\anaconda3\envs\mila\Lib\subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\terve\anaconda3\envs\mila\Lib\subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Users\terve\anaconda3\envs\mila\Lib\subprocess.py", line 1538, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [WinError 2] Le fichier spécifié est introuvable
```

### Screenshots
image (4)
image (3)
image (2)
image (1)
image

### Desktop (please complete the following information):
 - OS: [Windows 11]

### Additional context
I started following Victor Schmidt's guide (https://vsch.notion.site/YAMSS-5471da23464e41d4bad5e3517d273dea#0742904b6e384aba94b29a24b69e7b0e) on my WSL machine. But I decided to try milatools because I am on Windows, which is a problem if I want to open VS Code on a compute node following Victor's guide.

### LINKED ISSUE #66: The ControlMaster/ControlPath/ControlPersist SSH entries added by `mila init` are not supported on Windows
### Make sure you can reproduce the issue with the latest version available
```
pip install milatools --upgrade
mila init
```

### Describe the bug
Running `ssh mila` then outputs something like: `getsockname failed: Not a socket`.

https://github.com/microsoft/vscode-remote-release/issues/96

### LINKED ISSUE #69: The `hostname -s` command used to check if running on the Mila cluster isn't supported on Windows

### PR DESCRIPTION
**Pull request by lebrice**
- Fixes #63
- Fixes #66
- Fixes #69 
- Should hopefully fix #37 
- Should hopefully fix #19

### CODE REVIEW

File: milatools/cli/code_command.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import json
+from pathlib import Path
+from socket import timeout
```
Comment (breuleux) [inline milatools/cli/code_command.py:3]: Doesn't appear to be used.

File: milatools/cli/commands.py [hunk 2]
```diff
@@ -27,6 +27,8 @@
 from invoke import UnexpectedExit
 from typing_extensions import TypedDict
 
+from milatools.cli.code_command import set_remote_ssh_vscode_settings
```
Comment (breuleux) [inline milatools/cli/commands.py:30]: The other imports are relative, for consistency this one should be as well.

File: milatools/cli/commands.py [hunk 3]
```diff
@@ -418,7 +421,16 @@ def init():
...
         ):
-            here.run("ssh-copy-id", "mila")
+            # NOTE: If we're on a Windows machine, we do something different here:
+            if sys.platform == "win32":
+                command = (
+                    "powershell.exe type $env:USERPROFILE\\.ssh\\id_rsa.pub | ssh mila "
+                    '"cat >> ~/.ssh/authorized_keys"'
+                )
+                print(T.bold_green("(local) $ " + command))
+                subprocess.getoutput(command)
```
Comment (breuleux) [inline milatools/cli/commands.py:431]: Why not `here.run(command)`?

File: milatools/cli/commands.py [hunk 4]
```diff
@@ -561,15 +578,42 @@ def code(
...
     if not command_path:
         raise CommandNotFoundError(command)
+    qualified_node_name = qualified(node_name)
+
+    # TODO: Find the path to this file on other platforms.
+    vscode_settings_json_path = Path.home() / (
+        "AppData\\Roaming\\Code\\User\\settings.json"
+        if (sys.platform == "win32" or running_inside_WSL)
+        else "~/.config/Code/User/settings.json"
+    )
```
Comment (breuleux) [inline milatools/cli/commands.py:588]: The settings are in `$HOME/Library/Application\ Support/Code/User/settings.json` under MacOS.

File: milatools/cli/commands.py [hunk 5]
```diff
@@ -561,15 +578,42 @@ def code(
...
+
+    if vscode_settings_json_path.exists():
+        print(T.bold(f"Found VSCode settings file at {vscode_settings_json_path}"))
+        set_remote_ssh_vscode_settings(
+            vscode_settings_json_path=vscode_settings_json_path,
+            timeout_seconds=60,
+            fully_qualified_node_name=qualified_node_name,
+        )
```
Comment (breuleux) [inline milatools/cli/commands.py:596]: I feel like this should be in `mila init` instead and that there should be a confirmation of the diff by the user.

File: milatools/cli/code_command.py (hunk 1 above)
Comment (lebrice) [reply milatools/cli/code_command.py:3]: Yeah sorry this was still a bit of a WIP.

File: milatools/cli/commands.py (hunk 5 above)
Comment (lebrice) [reply milatools/cli/commands.py:596]: True!

File: milatools/cli/code_command.py (hunk 1 above)
Comment (lebrice) [reply milatools/cli/code_command.py]: Removed this for now in fc61b941fdd5b5ed907b1f667c24a5dd1cd20c10

File: milatools/cli/commands.py (hunk 4 above)
Comment (lebrice) [reply milatools/cli/commands.py]: Neat! I'll use that when I make a PR for that.
Removed this chunk for now in fc61b941fdd5b5ed907b1f667c24a5dd1cd20c10

File: milatools/cli/commands.py (hunk 3 above)
Comment (lebrice) [reply milatools/cli/commands.py:429]: I recall it not working for some reason. I'll try again on Windows and see if it works.

File: milatools/cli/commands.py (hunk 2 above)
Comment (lebrice) [reply milatools/cli/commands.py]: Fixed in fc61b941fdd5b5ed907b1f667c24a5dd1cd20c10

File: milatools/cli/commands.py (hunk 3 above)
Comment (lebrice) [reply milatools/cli/commands.py]: Fixed in 78fde43c3a89338f60ef0a2968e4b558132f2817

File: milatools/cli/commands.py (hunk 5 above)
Comment (lebrice) [reply milatools/cli/commands.py]: Removed this in fc61b941fdd5b5ed907b1f667c24a5dd1cd20c10

**Review by breuleux — APPROVED** [review]

**Comment by lebrice** [comment]
Hmm I just realized there's one last thing that might need fixing:
- If people run `mila init` inside WSL, then the ssh key setup is only done on the Ubuntu side, but the `code` command is actually run on the Windows side through Powershell!

**Comment by lebrice** [comment]
Ready for review again @breuleux

File: tests/cli/test_init_command/test_setup_windows_ssh_config_from_wsl_reject_.md
Comment (lebrice) [inline tests/cli/test_init_command/test_setup_windows_ssh_config_from_wsl_reject_.md:1]: Bug: This shouldn't be the case: We don't confirm the changes so the file shouldn't get copied.

Comment (lebrice) [reply tests/cli/test_init_command/test_setup_windows_ssh_config_from_wsl_reject_.md:1]: Fixed in 531b89a26cc093eb1cf1b303b59c753762d275be

**Review by satyaog — APPROVED** [review]

File: milatools/cli/init_command.py [hunk 6]
```diff
@@ -206,6 +261,7 @@ def _add_ssh_entry(
     ssh_config: SSHConfig,
     host: str,
     Host: str | None = None,
+    *,
```
[the remark this answers is not in the document]
Comment (lebrice) [reply milatools/cli/init_command.py:299]: It's making it so `_space_before` and `_space_after` cannot be passed as positional arguments. This helps remove a typing error when calls to `_add_ssh_entry` are made with different **kwargs (Windows vs non-Windows) in the next PRs.

[the remark this answers is not in the document]
Comment (satyaog) [reply milatools/cli/init_command.py:299]: ah get it thanks

[the remark this answers is not in the document]
Comment (satyaog) [reply milatools/cli/init_command.py:299]: Could we, should we assert that nothing is in this `*`?

```
def ... , *_args, ...:
    assert not _args
```

Only thinking out loud here

[the remark this answers is not in the document]
Comment (lebrice) [reply milatools/cli/init_command.py:299]: Exactly! That's what using a `*` without a name does

**Review by breuleux — APPROVED** [review]

**Pull request merged**